### PR TITLE
[TDW] move from npm to pnpm

### PR DESCRIPTION
**Pull request by segunadebayo**
# Description

This PR marks the start of improvements in our overall development experience. This will happen in two steps (and two pull requests):

- **Migration to PNPM (current):** This will reveal hoisted dependencies that weren't installed explicitly and could cause issues on the consumer side. It also helps rely less on turbo since pnpm has a task runner system (and it's fast)

- **Remove the need to run `watch` and read from source directly:** This will greatly improve TS performance and the need to "Restart TS Server" when working across packages. Of course, this will come in a different PR.

My end goal is to make our lives a bit easier. It's probably not much, but it's worth it.

For this PR, I've done the following:

- Moved workspace definition to `pnpm-workspace.yaml`
- Change all `@web3modal/*` package to point to `workspace:*`
- Renamed the commands across all packages to be shorter and compatible with the pnpm task runner
- (Debatable) Used pnpm task runner everywhere, it's almost as fast as turbo but without the caching (since we always use `cache: false` in turbo). Happy to bring back turbo once everything works as expected.

## Type of change

- [x] Chore (non-breaking change that addresses non-functional tasks, maintenance, or code quality improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Associated Issues

For Linear issues: Closes APKT-xxx
For GH issues: closes #...

# Showcase (Optional)

If there is a UI change include the screenshots with before and after state.
If new feature is being introduced, include the link to demo recording.

# Checklist

- [ ] Code in this PR is covered by automated tests (Unit tests, E2E tests)
- [ ] My changes generate no new warnings
- [ ] I have reviewed my own code
- [ ] I have filled out all required sections

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -1 +1,2 @@
 script-shell=sh
+shamefully-hoist=true
```
Comment (segunadebayo) [inline .npmrc:2]: This was the only way I could get `packages/cdn` to build. Without it, it throws some vite errors related to the node-polyfill.

Setting shamefully hoist makes it easy to use code from a dependency that has not been explicitly added to `package.json`

I'm unhappy with that, but I'll fix the cdn build in a separate PR/task.

File: Dockerfile.canary [hunk 2]
```diff
@@ -1,15 +1,19 @@
 FROM node:20-bookworm
+ENV PNPM_HOME="/pnpm"
+ENV PATH="$PNPM_HOME:$PATH"
+RUN corepack enable
```
Comment (segunadebayo) [inline Dockerfile.canary:4]: Got these from https://pnpm.io/docker

File: examples/vue-solana/src/App.vue [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <script lang="ts" setup>
 import {
   createWeb3Modal,
-  defaultConfig,
+  defaultSolanaConfig,
```
Comment (segunadebayo) [inline examples/vue-solana/src/App.vue:4]: There was no export of `defaultConfig` from the solane/vue package. The correct import is `defaultSolanaConfig`

Mostly mirrored the set up from `examples/react-solana`

File: package.json [hunk 4]
```diff
@@ -63,8 +44,10 @@
     "eslint-plugin-prettier": "5.1.3",
     "eslint-plugin-require-extensions": "0.1.3",
     "prettier": "3.1.1",
+    "test-exclude": "7.0.1",
```
Comment (segunadebayo) [inline package.json:47]: For some reason, `vitest` required this. I'm guessing it's related to the dependency hoisting again.

File: packages/solana/package.json [hunk 5]
```diff
@@ -19,6 +19,11 @@
       "import": "./dist/esm/exports/react.js",
       "default": "./dist/esm/exports/react.js"
     },
+    "./vue": {
+      "types": "./dist/types/exports/vue.d.ts",
+      "import": "./dist/esm/exports/vue.js",
+      "default": "./dist/esm/exports/vue.js"
+    },
```
Comment (segunadebayo) [inline packages/solana/package.json:26]: the `/vue` subpath was not exported. I wonder how this ever worked. The build failed without this.

File: Dockerfile.canary [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
+
 RUN apt update && apt install -y awscli
 
 WORKDIR /src
 
 COPY ../ ./
-RUN npm ci
-RUN npm run build
+RUN pnpm install
```
Comment (tomiir) [inline Dockerfile.canary:11]: should this be a clean install too?

File: .npmrc (hunk 1 above)
Comment (tomiir) [reply .npmrc:2]: @enesozturk @svenvoskamp any clues as to why this fails with the vite errors?

File: package.json [hunk 7]
```diff
@@ -4,54 +4,35 @@
...
+    "build:gallery": "pnpm build; pnpm --filter=./apps/gallery build",
+    "build:laboratory": "pnpm build; pnpm --filter=./apps/laboratory build",
+    "build:demo": "pnpm build; pnpm --filter=./apps/demo build",
+    "build:examples": "pnpm build; pnpm --filter=./examples -r build",
+    "test": "pnpm -r test --if-present",
     "coverage:merge": "node ./scripts/coverage-merge.js",
     "coverage:summary": "node ./scripts/coverage-summary.js",
-    "typecheck": "turbo run typecheck",
-    "lint": "turbo run lint",
+    "typecheck": "pnpm --if-present -r typecheck",
```
Comment (tomiir) [inline package.json:22]: why `--if-present` ?

File: packages/scaffold-utils/package.json [hunk 8]
```diff
@@ -33,15 +33,17 @@
...
+    "clean": "rm -rf dist",
+    "build": "tsc --build",
     "watch": "tsc --watch",
     "typecheck": "tsc --noEmit",
     "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
   },
   "dependencies": {
-    "@web3modal/core": "5.0.7",
-    "@web3modal/polyfills": "5.0.7",
+    "@coinbase/wallet-sdk": "4.0.3",
```
Comment (tomiir) [inline packages/scaffold-utils/package.json:43]: why add this dependency?

File: packages/solana/package.json (hunk 5 above)
Comment (tomiir) [reply packages/solana/package.json:26]: I believe this was recently merged by @phoenixVS

File: package.json (hunk 7 above)
Comment (segunadebayo) [reply package.json:22]: because not every package in the workspace has that command

File: packages/scaffold-utils/package.json (hunk 8 above)
Comment (segunadebayo) [reply packages/scaffold-utils/package.json:43]: https://github.com/WalletConnect/web3modal/blob/4a242fa2da1fef410e8df7c489204b0f3e4ad68c/packages/scaffold-utils/src/EthersTypesUtil.ts#L2

File: packages/solana/package.json (hunk 5 above)
Comment (segunadebayo) [reply packages/solana/package.json:26]: Great. I'll rebase accordingly

File: .npmrc (hunk 1 above)
Comment (segunadebayo) [reply .npmrc:2]: To be specific, here's the error

CleanShot 2024-07-18 at 17 41 46@2x

**Comment by lukaisailovic** [comment]
Changes look good.
Should we add a check in dangerfile if someone accidentally commit npm/yarn lock?

**Comment by segunadebayo** [comment]
@lukaisailovic I believe this is already covered here
CleanShot 2024-07-19 at 15 54 21@2x

File: Dockerfile.canary (hunk 6 above)
Comment (enesozturk) [reply Dockerfile.canary:11]: Wdym clean install? 🤔

File: apps/gallery/package.json [hunk 9]
```diff
@@ -4,14 +4,15 @@
...
-    "dev:gallery": "storybook dev -p 6006",
-    "build:gallery": "storybook build -o out",
+    "dev": "storybook dev -p 6006",
+    "build": "storybook build -o out",
     "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
   },
   "dependencies": {
-    "@web3modal/common": "5.0.7",
-    "@web3modal/ui": "5.0.7",
+    "@web3modal/common": "workspace:*",
```
Comment (enesozturk) [inline apps/gallery/package.json:12]: Finally. Love this feature.

File: package.json (hunk 7 above)
Comment (enesozturk) [reply package.json:22]: Shouldn't the turborepo already do this conditional run? If there is no script in a package, it's just not running as I know, prob something specific to pnpm?

File: package.json [hunk 10]
```diff
@@ -4,57 +4,39 @@
...
+    "build:laboratory": "pnpm build:packages; pnpm --filter=./apps/laboratory build",
+    "build:demo": "pnpm build:packages; pnpm --filter=./apps/demo build",
+    "build:examples": "pnpm build:packages; pnpm --filter=./examples -r build",
+    "test": "pnpm -r test --if-present",
     "coverage:merge": "node ./scripts/coverage-merge.js",
     "coverage:summary": "node ./scripts/coverage-summary.js",
-    "typecheck": "turbo run typecheck",
-    "lint": "turbo run lint",
+    "typecheck": "pnpm --if-present -r typecheck",
+    "lint": "pnpm -r lint",
```
Comment (enesozturk) [inline package.json:23]: [-r parameter](https://pnpm.io/cli/remove#--recursive--r) not really clear for me. Do we have to use it? It looks like we need to be careful about, could you explain why we used it?

Comment (segunadebayo) [reply package.json:23]: `--recursive` flag is used to run a script in every workspace package.

File: package.json (hunk 7 above)
Comment (segunadebayo) [reply package.json:22]: This is specific to `pnpm`

**Review by svenvoskamp — APPROVED** [review]
LGTM

**Review by tomiir — APPROVED** [review]

File: Dockerfile.canary (hunk 6 above)
Comment (tomiir) [reply Dockerfile.canary:11]: `npm ci` does a clean install in CI envs. Wondering if we need to do the same (-i think it's `pnpm --fronzen-lockfile`)

File: package.json (hunk 10 above)
Comment (enesozturk) [reply package.json:23]: Ah yes, the way they described is quite quite different just why:
> When used inside a [workspace](https://pnpm.io/workspaces), removes a dependency (or dependencies) from every workspace package.

**Pull request merged**